bpf: Handle fragments in SNAT flows

gentoo-root opened the pull request:
The code that extracts ports in snat_v4_nat and snat_v4_rev_nat doesn't take into account that the packet may be a second or further fragment. In this case, garbage will be read instead of the real port numbers. Fix this by using the existing ipv4_ct_extract_l4_ports that takes into account fragmentation.

This change makes the behavior of the switch inside the aforementioned functions closer to ct_extract_ports4.

Consider fragments in snat_v4_rewrite_ingress, avoid rewriting ports if it's not the first fragment.

Fixes: 14a653ad4aac ("bpf/nat: introduce snat_v4_nat() and snat_v4_rev_nat functions")

This should be applied on top of #25112.

```release-note
Handle IPv4 fragments in SNAT flows correctly.
```

gentoo-root committed:
bpf: Handle fragments in SNAT flows

The code that extracts ports in snat_v4_nat and snat_v4_rev_nat doesn't
take into account that the packet may be a second or further fragment.
In this case, garbage will be read instead of the real port numbers. Fix
this by using the existing ipv4_ct_extract_l4_ports that takes into
account fragmentation.

This change makes the behavior of the switch inside the aforementioned
functions closer to ct_extract_ports4.

Consider fragments in snat_v4_rewrite_ingress, avoid rewriting ports if
it's not the first fragment.

Fixes: 14a653ad4aac ("bpf/nat: introduce snat_v4_nat() and snat_v4_rev_nat functions")
Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Move fragment metric update to __ct_lookup4

The previous commit ("bpf: Handle fragments in SNAT flows") added an
extra call to ipv4_ct_extract_l4_ports into the SNAT paths. It can lead
to double-accounting of the same fragments, failing the "Supports IPv4
fragments" test, because ipv4_handle_fragmentation is called one more
time in those flows.

Fix it by moving the metric update out of ipv4_handle_fragmentation
directly to __ct_lookup4.

Suggested-by: Julian Wiedmann <jwi@isovalent.com>
Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1021,10 +1016,9 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target,
 	case IPPROTO_SCTP:
 #endif  /* ENABLE_SCTP */
-		if (ctx_load_bytes(ctx, off, &l4hdr, sizeof(l4hdr)) < 0)
+		if (ipv4_ct_extract_l4_ports(ctx, off, CT_INGRESS, &tuple, &has_l4_header) < 0)
```
Hm ideally we would pass the `ip4` through, and not re-validate it internally. But that's probably just a follow-on cleanup?

On the other hand could also just open-code the fragment look-up, same as `lb4_extract_tuple()`.

julianwiedmann replied:
Staged that cleanup here: https://github.com/cilium/cilium/pull/28717

I left it in draft for now, your call if we want to merge your PR first or rebase :).

gentoo-root replied:
My changes would fit naturally on top of your cleanup. Let's merge yours first.

julianwiedmann replied:
:merged:.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -491,43 +491,47 @@ static __always_inline int snat_v4_rewrite_ingress(struct __ctx_buff *ctx,
-			if (ret < 0)
-				return ret;
-			break;
+	if (has_l4_header) {
```
you beat me to converting this code path to `snat_v4_rewrite_headers()` :).

julianwiedmann commented on bpf/lib/conntrack.h:
```diff
@@ -873,9 +879,22 @@ ct_lazy_lookup4(const void *map, struct ipv4_ct_tuple *tuple,
+		return DROP_CT_INVALID_HDR;
+
+	is_fragment = ipv4_is_fragment(ip4);
+#endif
```
Could you easily pass a `is_fragment` parameter instead? Revalidating like this will access the wrong header if we're processing the inner payload of an encapsulated packet (ie. what we're doing for XDP-passthrough of GENEVE traffic, where `nodeport_lb4()` takes a `l3_off` != ETH_HLEN).

(uff ok, I had a quick look. For some of the callers that gets quite churny ...)

gentoo-root replied:
> Revalidating like this will access the wrong header

Is it only relevant for ct_lazy_lookup4 flows? For ct_lookup4 such a revalidation used to happen before my patch as well: ct_lookup4 -> ct_extract_ports4 -> **ipv4_ct_extract_l4_ports** -> ipv4_handle_fragmentation -> ipv4_is_fragment. Is it also buggy?

julianwiedmann replied:
Yep, right now it's only a concern for some of the `ct_lazy_lookup4()` calls in `nodeport_lb4()`, when used from XDP.

gentoo-root replied:
I'll do this change. It seems, however, that just passing is_fragment = false would have the same behavior as before this patch.

julianwiedmann replied:
looks like this is becoming a headache :/.

Your idea would be to pass `is_fragment = false` for `ct_lazy_lookup4()` users, and then convert them over 1-by-1 as needed? I think that would be ok, and allow us to make progress on this massive improvement :+1:.

julianwiedmann reviewed (changes requested):
Thanks for tackling this! It's been a long time coming :).

A few smaller notes, but the only thing that really needs addressing is how `ct_lazy_lookup4()` obtains the `is_fragment` variable. Happy to brain-storm further if passing around an additional parameter is too much hassle.

gentoo-root committed:
bpf: Use snat_v4_rewrite_headers for snat_v4_rev_nat

Replace snat_v4_rewrite_ingress by a generic snat_v4_rewrite_headers for
better code reuse. This change should not change the behavior, the new
code should be functionally equivalent to the old one.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -393,25 +390,37 @@ snat_v4_rewrite_headers(struct __ctx_buff *ctx, __u8 nexthdr, int l3_off,
-			if (l4_store_port(ctx, l4_off, port_off, new_port) < 0)
-				return DROP_WRITE_ERROR;
+			/* Amend the L4 checksum due to changing the ports. The L3 checksum stays
+			 * the same because changing both the port and the L4 checksum compensate
```
This I don't quite parse - the IPv4 header checksum shouldn't even cover the L4 header (or any change to its port fields) ?

gentoo-root replied:
Ah right, I got confused when I wrote this comment, let me fix it.

It's relevant for a whole-packet checksum (such as skb->csum for CHECKSUM_COMPLETE), but indeed the IPv4 checksum covers only the headers.

julianwiedmann reviewed (changes requested):
Thank you! Looks good, just the `ct_lazy_lookup4()` that needs adjusting.

Should we maybe expedite the checksum fixes through a separate PR ?

gentoo-root committed:
bpf: Pass is_fragment to ct_lazy_lookup4

The next commit will use it to update metrics. Note that revalidate_data
shouldn't be done in place, because the L3 header offset may be not
equal to ETH_HLEN in XDP flows.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2895,7 +2895,7 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 		__ipv4_ct_tuple_reverse(&tuple);
 
-		ret = ct_lazy_lookup4(get_ct_map4(&tuple), &tuple, ctx, l4_off, has_l4_header,
+		ret = ct_lazy_lookup4(get_ct_map4(&tuple), &tuple, ctx, ip4, l4_off, has_l4_header,
```
Not sure if you've seen the verifier reject in CI. But it's almost certainly complaining about this occurence, as the rewrites in `lb4_local() ` will invalidate the pointer.

gentoo-root replied:
Oh, this is why it fails =/

Thanks, I didn't realize it was a verifier error. Could you point me where you see it?

julianwiedmann replied:
I just randomly poked around in the red CI results, saw the error message in https://github.com/cilium/cilium/actions/runs/6696314287/job/18194555181?pr=25340 and had a look at the agent log in the sysdump.

gentoo-root replied:
Hmm, weird that BPF Checks didn't turn red with the relevant verifier error in the action log.

gentoo-root replied:
Phew, I think I patched that. CI is green.

julianwiedmann reviewed (approved):
ty Maxim, looks great!
